### PR TITLE
Spanner hf transaction lab lock

### PR DESCRIPTION
**Pull request by bajpaipankaj28**
This PR contains the logic to reproduce the Locking and Non locking behavior in Spanner. Locking method is using the wrong implementation which will produces lock and other issues the way code is written. Unlocking method will help in showing the right way of using it.

In table we have used generated_value  which is string and UUID value. This is a non index column so running a query even on single row of it will be the full table scan.

### CODE REVIEW

File: samples/snippets/src/main/java/com/example/spanner/SpannerSample.java [hunk 1]
```diff
@@ -1956,14 +1954,17 @@ static void run(
           hubbleTransactionsCodeLab.createWorkItems(dbAdminClient, database);
 	break;
       case "transactionLabWriteWorkItems":
-          hubbleTransactionsCodeLab.writeWorkItems(dbClient, 10000, 50);
+          hubbleTransactionsCodeLab.writeWorkItems(dbClient, 1000, 50);
```
Comment (bajpaipankaj28) [inline samples/snippets/src/main/java/com/example/spanner/SpannerSample.java:1957]: With 1000 locking method was taking 15 min, so I reduced it to 1k from 10k.

File: samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java [hunk 2]
```diff
@@ -290,6 +242,175 @@ public void updatesAndReads(
...
   }
 
+  public void doWorkSingleTransactionParallelLocking(DatabaseClient dbClient) {
+    // getting the available processor
+    ExecutorService executorService = Executors.newFixedThreadPool(NUM_PROCESSORS);
+
+    List<Future<Integer>> futures = new ArrayList<>();
+    // List to store the count of total number of workItem(row) which has not yet completed means
+    // is_done is false.
+    List<String> workItemCount = new ArrayList<>();
```
Comment (nginsberg-google) [inline samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java]: We don't need this. Instead, we can just inspect the resultSet returned from the executed query. If there are zero rows in it, we know we've done all the work.

File: samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java [hunk 3]
```diff
@@ -290,6 +242,175 @@ public void updatesAndReads(
...
+      // Thread 2: ThreadIndex(i) 1  workitem 30
+      // Thread 3: ThreadIndex(i) 2  workitem 30
+      // ...
+      // Thread 10: ThreadIndex(i) 9  workitem 30
+      // These values will be determined by received by each thread(an object of WorkItemProcessor)
+      // while executing based on the input values passed here.
+      String nonLockQueryInTransaction =
+          "SELECT * FROM WorkList WHERE is_done is false and id= '%s';";
+      Callable<Integer> callable =
+          new WorkItemProcessor<>(
```
Comment (nginsberg-google) [inline samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java]: This is a lot of indirection. I would prefer that we write these as one function and get rid of the WorkItemProcessor class.

File: samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java [hunk 4]
```diff
@@ -290,6 +242,175 @@ public void updatesAndReads(
...
+
+      // Thread 1: ThreadIndex(i) 0, workitem 30
+      // Thread 2: ThreadIndex(i) 1  workitem 30
+      // Thread 3: ThreadIndex(i) 2  workitem 30
+      // ...
+      // Thread 10: ThreadIndex(i) 9  workitem 30
+      // These values will be determined by received by each thread(an object of WorkItemProcessor)
+      // while executing based on the input values passed here.
+      String lockQueryInTransaction =
+          "SELECT * FROM WorkList WHERE is_done is false and generated_value= '%s';";
```
Comment (nginsberg-google) [inline samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java]: The SQL command that you should run is "SELECT * FROM WorkList WHERE is_done is false LIMIT 100", and then use a random number to pick from the rows returned which WorkItem to do. This should let you remove generated_value, as well as remove the precompute step where you store all the workitems locally.

File: samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java [hunk 5]
```diff
@@ -290,6 +242,175 @@ public void updatesAndReads(
...
+    // This will help us to assign executing the random workItem(rows) work without overlapping. As
+    // each thread will pick up different generated_value to query.
+    try (ResultSet resultSet =
+        dbClient
+            .singleUse() // Execute a single read or query against Cloud Spanner.
+            // Below query can be parameterized to but kept here for simplicity
+            .executeQuery(
+                Statement.of("SELECT generated_value FROM WorkList WHERE is_done is false"))) {
+      while (resultSet.next()) {
+        workItemCount.add(resultSet.getString(0));
```
Comment (nginsberg-google) [inline samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java]: This, in general, is not a pattern we should use. What if the workItem set is being updated? What if the workItem set doesn't fit into local memory? While this works in our toy example, it is very unlikely to work in a real world use case.

File: samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java [hunk 6]
```diff
@@ -290,6 +242,175 @@ public void updatesAndReads(
...
+        totalProcessed += future.get();
+      }
+      executorService.shutdown();
+      executorService.awaitTermination(Long.MAX_VALUE, TimeUnit.NANOSECONDS);
+    } catch (InterruptedException | ExecutionException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public void writeMessagesInterleavedParallel100k(
```
Comment (nginsberg-google) [inline samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java]: Please put the writeMessages functions into a separate PR.

File: samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java [hunk 7]
```diff
@@ -290,6 +242,175 @@ public void updatesAndReads(
...
+    // getting the available processor
+    ExecutorService executorService = Executors.newFixedThreadPool(NUM_PROCESSORS);
+
+    List<Future<Integer>> futures = new ArrayList<>();
+    // This is the store the count of total number of workItem(row) which has not yet completed
+    // means  is_done is false.
+    List<String> workItemCount = new ArrayList<>();
+    // Below logic is reading the id(primary key) for the rows which are having is_done=false. This
+    // will help us to read the different ids by different thread randomly.
+    try (ResultSet resultSet =
```
Comment (nginsberg-google) [inline samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java]: Same comment as above - this is not a realistic pattern.

File: samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java [hunk 8]
```diff
@@ -290,6 +242,175 @@ public void updatesAndReads(
...
+      // E.g In case of 10 threads and 30 workItem the assignment for this method would be like
+
+      // Thread 1: ThreadIndex(i) 0, workitem 30
+      // Thread 2: ThreadIndex(i) 1  workitem 30
+      // Thread 3: ThreadIndex(i) 2  workitem 30
+      // ...
+      // Thread 10: ThreadIndex(i) 9  workitem 30
+      // These values will be determined by received by each thread(an object of WorkItemProcessor)
+      // while executing based on the input values passed here.
+      String nonLockQueryInTransaction =
```
Comment (nginsberg-google) [inline samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java]: What we need to do here is:

singleUse transaction of "SELECT id FROM WorkList WHERE is_done is false LIMIT 100"
random number to pick from the returned rows

readwrite transaction that re-reads the selected row to ensure that the work was not done by another thread
updates the work item

This will let you remove the precompute step, and will also mean that you can probably no longer use WorkItemProcessor.

File: samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java [hunk 9]
```diff
@@ -392,4 +513,112 @@ private void executeTasksInParallel(Runnable task, int numProcessors) {
       executorService.shutdown();
     }
   }
+
+  /**
+   * Represents a task to process workItem(rows) in parallel, distributing work among threads based
+   * on a specific value.
+   */
+  static class WorkItemProcessor<T> implements Callable<Integer> {
+    private final List<T> totalWorkItem;
```
Comment (nginsberg-google) [inline samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java]: Instead of precomputing a list of work items per thread, we can just have all the threads run until they cannot find an undone work item.

**Review by nginsberg-google — CHANGES_REQUESTED** [review]

File: samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java (hunk 4 above)
Comment (nginsberg-google) [reply samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java]: (this query should run as part of the readwrite transaction)

File: samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java (hunk 2 above)
Comment (bajpaipankaj28) [reply samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java]: Here do ups mean result.getRows()--> I did not find this method in the API? or something else

File: samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java (hunk 4 above)
Comment (bajpaipankaj28) [reply samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java]: What I understand is Query-> "SELECT * FROM WorkList WHERE is_done is false LIMIT 100" will run outside of any transaction(the way I ran above one but in on method) , then pick a row randomly out of those 100 and use a query to fetch data for that specific row in ReadwriteTransaction(). If **is_done** is still false update it in same transaction else exit and pick up any thing new.

Comment (nginsberg-google) [reply samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java]: This is true for the second "correct" method. For the first method, it should run as part of the read/write transaction.

File: samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java (hunk 2 above)
Comment (nginsberg-google) [reply samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java]: result.getRows().size() != 0

File: samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java [hunk 10]
```diff
@@ -87,55 +87,102 @@ public void writeWorkItems(
...
   }
 
-  public void doWorkSingleTransactionSerial(DatabaseClient dbClient, boolean doneValue) {
+  private void doWorkSingleTransactionSerial(
+      DatabaseClient dbClient, boolean doneValue, boolean isLocking) {
     boolean workRemaining = true;
     int workDone = 0;
     while (workRemaining) {
-      workRemaining = doWorkSingleTransaction(dbClient, doneValue);
-      if (workDone++ % 1000 == 0) {
```
Comment (bajpaipankaj28) [inline samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java:120]: As this count will be used by individual thread so it will not reach 1000, it may be different for different thread. Hence I remove the print statement. If we really need to print something like it we may need add a logic or if printing individual thread is fine then I can add something here it self.

File: samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java [hunk 11]
```diff
@@ -87,55 +87,102 @@ public void writeWorkItems(
...
-        System.out.printf(">>>>>>>>>>>>>>>> Done %d work <<<<<<<<<<<<<\n", workDone);
-      }
+      workRemaining =
+          isLocking
```
Comment (bajpaipankaj28) [inline samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java:210]: Using this we can leverage one method, we can create duplicate method if its not expected this way.

File: samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java [hunk 12]
```diff
@@ -87,55 +87,102 @@ public void writeWorkItems(
     }
   }
 
-  public boolean doWorkSingleTransaction(DatabaseClient dbClient, boolean doneValue) {
+  private boolean doWorkSingleTransactionLocking(DatabaseClient dbClient, boolean doneValue) {
```
Comment (nginsberg-google) [inline samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java]: I think we can remove `doneValue`, and just use false. I included it so that as I was debugging this I could move the worklists back and forth between true and false without creating a new table, but that is unnecessary for the codelab.

File: samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java [hunk 13]
```diff
@@ -87,55 +87,102 @@ public void writeWorkItems(
...
+
+    String stmt = String.format("SELECT * FROM WorkList WHERE is_done is %b LIMIT 100", !doneValue);
+
+    return Boolean.TRUE.equals(
```
Comment (nginsberg-google) [inline samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java:101]: Let's add some comments please

File: samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java [hunk 14]
```diff
@@ -87,55 +87,102 @@ public void writeWorkItems(
...
+    }
+    if (idList.size() == 0) return false;
+
+    return Boolean.TRUE.equals(
+        dbClient
+            .readWriteTransaction()
+            .run(
+                transaction -> {
+                  Random random = new Random();
+                  int randIdx = random.nextInt(idList.size());
```
Comment (nginsberg-google) [inline samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java]: You can compute this outside of the transaction.

File: samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java [hunk 15]
```diff
@@ -87,55 +87,102 @@ public void writeWorkItems(
...
+                  String id = idList.get(randIdx);
+                  Struct idRow =
```
Comment (nginsberg-google) [inline samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java:180]: This should be the first line of the transaction.

File: samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java [hunk 16]
```diff
@@ -87,55 +87,102 @@ public void writeWorkItems(
...
+                      transaction.readRow(
+                          "WorkList",
+                          Key.of(id), // Read single row in a table.
+                          Arrays.asList("id", "is_done", "timestamp"));
+
+                  if (idRow == null || idRow.getBoolean("is_done")) return false;
```
Comment (nginsberg-google) [inline samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java]: Should idRow == null ever happen? Is that an error?

File: samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java (hunk 10 above)
Comment (nginsberg-google) [reply samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java:120]: This was a debugging statement, totally fine to remove :)

**Review by nginsberg-google — CHANGES_REQUESTED** [review]

File: samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java [hunk 17]
```diff
@@ -87,102 +87,136 @@ public void writeWorkItems(
...
         dbClient
+            // In this logic read and write is happening in same transaction.
             .readWriteTransaction()
             .run(
                 transaction -> {
-                  Random random = new Random();
                   List<String> idList = new ArrayList<>();
+                  // Executing the above query to get 100 row where is_done is false, it will result
+                  // a
+                  // full table scan
```
Comment (nginsberg-google) [inline samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java]: Can this be on the same line as above?

File: samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java [hunk 18]
```diff
@@ -87,102 +87,136 @@ public void writeWorkItems(
...
                           .set("timestamp")
                           .to(Value.COMMIT_TIMESTAMP)
                           .build());
                   return true;
                 }));
   }
 
-  private boolean doWorkSingleTransactionNonLocking(DatabaseClient dbClient, boolean doneValue) {
+  /*
+  This method helps reproduce the non Locking scenario, its the correct way using read and write in Spanner. This method
```
Comment (nginsberg-google) [inline samples/snippets/src/main/java/com/example/spanner/HubbleTransactionsCodeLab.java]: nit: it's

**Review by nginsberg-google — APPROVED** [review]
LGTM! Thanks for all the hard work on this, really appreciate it. Just a few small comments about formatting, but feel free to merge without another review.

**Comment by bajpaipankaj28** [comment]
Thank you so much for reviewing it patiently.
